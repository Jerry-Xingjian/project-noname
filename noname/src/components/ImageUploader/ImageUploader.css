#add:hover {
	cursor: pointer;
}

#add {
	position: relative;
	left: 20%;
}