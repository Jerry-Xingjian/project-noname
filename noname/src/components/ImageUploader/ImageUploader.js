import React from 'react';
import ImageUploading from 'react-images-uploading';
import uuid from 'react-uuid';
import { Alert, Button, ButtonGroup } from 'reactstrap';
import './ImageUploader.css';

function ImageUploader(props) {
  const {
    existingImages = [],
    maxNumber = 10,
    acceptType = ['jpeg', 'jpg', 'png', 'gif', 'mp4', 'mov', 'avi', 'wmv'],
    maxFileSize = 5000000,
    onImagesUpload,
  } = props;

  const [images, setImages] = React.useState(existingImages);

  const onChange = (imageList) => {
    setImages(imageList);
  };

  const onError = () => {
    setImages([]);
  };

  return (
    <div className="App">
      <ImageUploading
        multiple
        value={images}
        onChange={onChange}
        onError={onError}
        maxNumber={maxNumber}
        acceptType={acceptType}
        maxFileSize={maxFileSize}
        dataURLKey="url"
      >
        {({
          imageList,
          onImageUpload,
          onImageRemoveAll,
          onImageUpdate,
          onImageRemove,
          errors,
        }) => (
          <>
            {errors && (
              <Alert color="danger text-start">
                <ul>
                  {errors.maxNumber && (
                    <li>Number of selected images exceed maxNumber</li>
                  )}
                  {errors.acceptType && (
                    <li>Your selected file type is not allow</li>
                  )}
                  {errors.maxFileSize && (
                    <li>Selected file size exceed maxFileSize</li>
                  )}
                </ul>
              </Alert>
            )}

            <div className="upload__image-wrapper align-items-center">
              <div className="p-2" style={{ textAlign: 'left' }}>
                {imageList.map((image, index) => (
                  <div
                    key={uuid()}
                    className="image-item"
                    style={{
                      width: '150px',
                      marginRight: '10px',
                      display: 'inline-block',
                    }}
                  >
                    <img
                      src={(image.url === undefined)
                        ? image
                        : image.url}
                      alt=""
                      style={{ width: '100%' }}
                    />
                    <div className="image-item__btn-wrapper mt-1">
                      <ButtonGroup size="sm" style={{ width: '100%' }}>
                        <Button
                          color="primary"
                          onClick={() => onImageUpdate(index)}
                        >
                          Update
                        </Button>
                        <Button
                          color="danger"
                          onClick={() => onImageRemove(index)}
                        >
                          Remove
                        </Button>
                      </ButtonGroup>
                    </div>
                  </div>
                ))}
              </div>
              <div
                className="container"
                id="add"
                role="presentation"
                onClick={onImageUpload}
                onKeyDown={onImageUpload}
              >
                <img src="https://nonome-project-media.s3.amazonaws.com/icons/directbox-send.svg" className="mx-2" alt="+" width="20" height="20" />
                Upload Photo Here
              </div>
              {images.length > 0 && (
                <>
                  <hr />
                  <div className="text-start p-2">
                    <Button onClick={() => onImagesUpload(images)} color="success">
                      Upload
                    </Button>
                    {' '}
                    <Button onClick={onImageRemoveAll} color="danger">
                      Remove All Images
                    </Button>
                  </div>
                  <pre className="text-start" id="jsonprint" />
                </>
              )}
            </div>
          </>
        )}
      </ImageUploading>
    </div>
  );
}

export default ImageUploader;
