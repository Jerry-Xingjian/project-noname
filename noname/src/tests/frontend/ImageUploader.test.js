import React from "react";
import '@testing-library/jest-dom';
import ImageUploader from "../../components/ImageUploader/ImageUploader";
import renderer from 'react-test-renderer';
import { render, screen, waitFor, fireEvent } from "@testing-library/react";

describe("Snapshot test", () => {
  test("ImageUploader matches Snapshot", async () => {
    const component = renderer.create(
      <ImageUploader />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Component test", () => {
  test("check whether certain text exits in document", async() => {
    render(
      <ImageUploader />
    );
    const uploadText = screen.getByText('Upload Photo Here');
    expect(uploadText).toBeInTheDocument()
  });
})

describe("Event test", () => {
  let file;

  beforeEach(() => {
    file = new File(["(⌐□_□)"], "test.png", { type: "image/png" });
  });

  test("image upload test", async() => {
    render(
      <ImageUploader />
    );
    // const uploadText = screen.getByText('Upload Photo Here');
    const uploader = screen.getByRole('presentation');

    await waitFor(() =>
      fireEvent.click(uploader, {
        target: { files: [file] },
      })
    );

    let image = document.getElementById("add");
    expect(image.files[0].name).toBe("test.png");
    expect(image.files.length).toBe(1);
  })
})

// todo: after upload, how to test the rerendered DOM