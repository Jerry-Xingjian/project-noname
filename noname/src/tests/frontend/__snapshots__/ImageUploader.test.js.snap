// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test ImageUploader matches Snapshot 1`] = `
<div
  className="App"
>
  <input
    accept=".jpeg, .jpg, .png, .gif, .mp4, .mov, .avi, .wmv"
    multiple={true}
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
  <div
    className="upload__image-wrapper align-items-center"
  >
    <div
      className="p-2"
      style={
        Object {
          "textAlign": "left",
        }
      }
    />
    <div
      className="container"
      id="add"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="presentation"
    >
      <img
        alt="+"
        className="mx-2"
        height="20"
        src="https://nonome-project-media.s3.amazonaws.com/icons/directbox-send.svg"
        width="20"
      />
      Upload Photo Here
    </div>
  </div>
</div>
`;
